src/Build/Logging/BinaryLogger/Postprocessing/TransparentReadStream.cs

```diff
@@ -0,0 +1,122 @@
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices.ComTypes;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Logging
+{
+    /// <summary>
+    /// A wrapper stream that allows position tracking and forward seeking.
+    /// </summary>
+    internal class TransparentReadStream : Stream
+    {
+        private readonly Stream _stream;
+        private long _position;
+
+        public static Stream EnsureSeekableStream(Stream stream)
+        {
+            if (stream.CanSeek)
+            {
+                return stream;
+            }
+
+            if(!stream.CanRead)
+            {
+                throw new InvalidOperationException(ResourceUtilities.GetResourceString("Binlog_StreamUtils_MustBeReadable"));
+            }
+
+            return new TransparentReadStream(stream);
+        }
+
+        public static TransparentReadStream EnsureTransparentReadStream(Stream stream)
+        {
+            if (stream is TransparentReadStream transparentReadStream)
+            {
+                return transparentReadStream;
+            }
+
+            if (!stream.CanRead)
+            {
+                throw new InvalidOperationException(ResourceUtilities.GetResourceString("Binlog_StreamUtils_MustBeReadable"));
+            }
+
+            return new TransparentReadStream(stream);
+        }
+
+        private TransparentReadStream(Stream stream)
+        {
+            _stream = stream;
+        }
+
+        public int? BytesCountAllowedToRead
+        {
+            set { _maxAllowedPosition = value.HasValue ? _position + value.Value : long.MaxValue; }
+        }
+
+        // if we haven't constrained the allowed read size - do not report it being unfinished either.
+        public int BytesCountAllowedToReadRemaining =>
+            _maxAllowedPosition == long.MaxValue ? 0 : (int)(_maxAllowedPosition - _position);
+
+        private long _maxAllowedPosition = long.MaxValue;
+        public override bool CanRead => _stream.CanRead;
+        public override bool CanSeek => true;
+        public override bool CanWrite => false;
+        public override long Length => _stream.Length;
+        public override long Position
+        {
+            get => _position;
+            set => this.SkipBytes((int)(value - _position), true);
+        }
+
+        public override void Flush()
+        {
+            _stream.Flush();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (_position + count > _maxAllowedPosition)
+            {
+                throw new StreamChunkOverReadException(
+                    ResourceUtilities.FormatResourceStringStripCodeAndKeyword("Binlog_StreamUtils_OverRead", count,
+                        _maxAllowedPosition - _position));
```

Comment:
This is the most interresting point from the review.
I still believe we want this and it's a legitimate step outside the general baseclass documented use.

**The motivation:**
tl;dr;: this - https://github.com/dotnet/msbuild/pull/9307/files#diff-3ddd8d601460ba4c5ec2996442f059a2e7c2659c00ac7acd346931434722faddR225
One of the 'forward compatibility reading' scenarios I want to support is a future change in internal representation of pre-existing BuildEventArgs data type, while such change is not append-only.
It's possible to detect such change and gracefully point it out and even recover if wished by caller. That can lead to better user experience. One subscenario is when members are removed (or their type changed) from a type in a future revisions. The reader code would then try to read extra data from the `StreamReader` and the easiest way to detect this is by throwing in such case. This is the easiest way of surfacing the broken expectation to the `BuildEventArgsReader`, without the need to cache the last read data (to be able to 'rollback' the over-read attempt).
Also to be able to recover from over-read attempt, we need to break the attempt before it is performed - otherwise we would lose the data that instead belongs to the next record.

**Possible Side Effects:**
I believe there are none - as we allways read through `BinaryReader` through which we allways try to read exact amount of bytes (for a given primitive type or for a given size of string).
Also the behavior is switched on only when explicitly opted in on that stream: https://github.com/dotnet/msbuild/pull/9307/files#diff-844cbc09b4b4aa566826f89cb133e56efc10a7ee54f85f0a98b7d632d5f001b1R59-R62
The opting-in property should however be very clearly documented (which it is not today).

**Alternative:**
 * Not supporting detection of over-reading
 * Implementing custom rollback capable `Stream` or `BinaryReader` - those would just rise some flag after over-reads and then would be able to replay the performed read as if it did not happen. This would require extra caching for each read call.

**Thoughts?**
Does the behavior sound acceptable under the presented conditions? Or would you suggest different choice? (roll back capable caching stream? Any other way?)

Reply:
There is one more alternative - replacing the `throw` code with:

```csharp
count = (int)(_maxAllowedPosition - _position);
```

This will evectively lead to `EndOfStreamException` and rest of logic remains identical (just the catch need to adjust).

Slight disadvantage is that we cannot pass any custom information with the exception and that we're not able to distinguish true end of stream (corrupted log) and end of substream. Though the true end of stream would trip on the attempt to read the next record type.

**Does this sound as the proper replacement?**

Reply:
Yes, adjusting `count` instead of throwing sounds like a good change to me.

> One subscenario is when members are removed (or their type changed) from a type in a future revisions.

My first reaction to this is: Should be super rare and if we really have a good reason to make such changes/overhaul, we would bump min reader version.

Reply:
Understood.

I agree in the first part (conforming to Stream contracts).

I disagree with the second part (event format change should lead to reader version increase). The reason being that reader is able to recover gracefully and the API user have the last call in deciding whether such change is breaking for them or not (discussed in the suggested doc: https://github.com/dotnet/msbuild/pull/9307/files#diff-10da6addf61b7431a6e2deb63435f116d5402ebb0532b5ce3c97b54a0438f721R138-R140).

E.g. if I'd be writing analysis tools that extracts and analyses all errors and warnings - I do not care that `AssemblyLoadedBuildEventArgs` changed its format. Would I have been writting the tool listing all assemblies loaded during the build - it would be breaking for me (but I wouldn't mind if I cannot properly read `BuildWarningEventArgs`).

The required reader version should ideally only idnicate major structural changes that would make reader unusable.

Reply:
> E.g. if I'd be writing analysis tools that extracts and analyses all errors and warnings - I do not care that AssemblyLoadedBuildEventArgs changed its format. 

That's a fair point. Although if I have a tool that analyses `AssemblyLoadedBuildEventArgs` and we allow arbitrary modifications to the structure, throwing on an over-read is not guaranteed to help. If we remove a field in one version and then add another field in the next version, the length may be the same but my tool would be reading bogus data.. Basically, my point is that if we really want to allow arbitrary changes and still be forward compatible, we would need the file to include a schema of some sort. Without it it's still best effort (and in this case also against an API semantic convention).

Reply:
Great point on theoretical possibility of colision of format of 2 different versions!
The documentation needs to be very clear about this.

I suggest:
 * [will update] Let's update the doc to explicitly call out it's a best effort compatibility and refrences update (to match not only the reader version, but the binlog version as well) is allways preferable option.
 * [stays as is] MSBuild devs can decide per each OM change if they want to increase required reader version or not.
 * [stays as is] API users can decide whether format error is breaking even if reader version supports it.

Does that sound acceptable?

Reply:
Sounds good, thank you.

Reply:
Documentation updated.
Plus called out the specific case you brought: https://github.com/dotnet/msbuild/pull/9307/files#diff-10da6addf61b7431a6e2deb63435f116d5402ebb0532b5ce3c97b54a0438f721R170